### ISSUE TITLE
WebSessionContext in ASP.NET Core 2

### ISSUE THREAD

**Issue by leandro86** [nhibernate/nhibernate-core#1632]
Hello,

I'm trying to use the WebSessionContext in a ASP.NET Core 2 application, but I get the following error when CurrentSessionContext is accesed:

> Could not load file or assembly 'System.Web, Version=4.0.30319.42000, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.

The problem seems to be in ReflectiveHttpContext.cs:

```c#
private static System.Type HttpContextType
{
	get
	{
		return
			System.Type.GetType(
				string.Format(
					"System.Web.HttpContext, System.Web, Version={0}, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
					Environment.Version));
	}
}
```

HttpContext is now in the assembly Microsoft.AspNetCore.Http.Abstractions. However, I don't believe it would be possible to load the current HttpContext by reflection, since now it doesn't have a "Current" property. ReflectiveHttpContext access the "Current" property of HttpContext here:

```c#
private static void CreateCurrentHttpContextGetter()
{
	PropertyInfo currentProperty = HttpContextType.GetProperty("Current", BindingFlags.Static | BindingFlags.Public | BindingFlags.FlattenHierarchy);
	Expression propertyExpression = Expression.Property(null, currentProperty);
	Expression convertedExpression = Expression.Convert(propertyExpression, typeof (object));
	HttpContextCurrentGetter = (Func<object>) Expression.Lambda(convertedExpression).Compile();		
}
```

The current HttpContext is now available through IHttpContextAccesor. See: https://stackoverflow.com/questions/31243068/access-the-current-httpcontext-in-asp-net-core

**Comment by fredericDelaporte** [comment]
Feature parity between full Framework/.Net Core about the `HttpContext` based session context seems hard to achieve, since the Core team is banning static accessors and tries hard to prevent/dissuade people of putting workarounds in place.

The main difficulties lie in the fact that NHibernate has been ported to .Net Core and not rewritten/re-architectured for it. It especially has not been rewritten as an injectable .Net Core service, meaning it has no mechanism out of the box for easily getting injected with other .Net Core services, like an `IHttpContextAccesor`. (Moreover this service is not guaranteed to be present. And many NHibernate setup may not require it either, so putting it as a mandatory dependency of NHibernate does not look as an option to me.)

I think this case would require a new kind of session context, which would also be a Core service to add in the pipeline like any service, allowing it to mandate and get any .Net Core dependencies it requires.

**Comment by leandro86** [comment]
Thanks for your answer. As a workaround, right now I'm doing this:

I created a new session context:

```c#
[Serializable]
public class AspNetCoreWebSessionContext : MapBasedSessionContext
{
	private const string SessionFactoryMapKey = "NHibernate.Context.AspNetCoreWebSessionContext.SessionFactoryMapKey";

	public static IHttpContextAccessor HttpContextAccessor { get; set; }

	public AspNetCoreWebSessionContext(ISessionFactoryImplementor factory) : base(factory)
	{
	}

	protected override IDictionary GetMap()
	{
		return HttpContextAccessor.HttpContext.Items[SessionFactoryMapKey] as IDictionary;
	}

	protected override void SetMap(IDictionary value)
	{
		HttpContextAccessor.HttpContext.Items[SessionFactoryMapKey] = value;
	}
}
```

And then, after building the session factory, I ask the container to resolve an instance of IHttpContextAccesor:

```c#
var sessionFactory = config.BuildSessionFactory();
AspNetCoreWebSessionContext.HttpContextAccessor = container.GetInstance<IHttpContextAccessor>();
```

It's also possible to create an instance of HttpContextAccessor directly inside AspNetCoreWebSessionContext, but that would require a dependency on Microsoft.AspNetCore.Http, instead of Microsoft.AspNetCore.Http.Abstractions.

What do you have in mind to do this the proper way?

**Comment by fredericDelaporte** [comment]
As a workaround you may use instead the `AsyncLocalSessionContext`, it should be suitable.

About providing a .Net Core compatible web session context, I would need to investigate the subject more.

**Comment by jpenniman** [comment]
In our  ASP.Net Core apps, we don't use the `WebSessionContext` at all. We manage the `ISessionFactory` and `ISession` life cycle with the dependency injection built into ASP.Net Core...

``` cs
services.AddSingleton<ISessionFactory>((provider)=>{ 
    var cfg = new NHibernate.Cfg.Configuration();
    // your config stuff here
    return cfg.BuildSessionFactory();
});

services.AddScoped<ISession>((provider)=>{
    var factory = provider.GetService<ISessionFactory>();
    return factory.OpenSession();
});
```

Then just inject `ISession` into what ever needs it.  For example (over simplified for brevity):

``` cs
public class CustomerController : Controller
{
    ISession session;

    public CustomerController(ISession session)
    {
        this.session = session;
    }

    public IEnumerable<Customer> Get()
    {
        return session.QueryOver<Customer>().List();
    }
}
```

We simplified it even further and created an extension method on `IServiceProvider` to handle all the boilerplate, so the Startup.cs in our projects just contains...

``` cs
services.AddNHibernate();
```

We have been using this approach for the last couple years (since ASP.Net Core 1.0) and it has worked well for us.

**Comment by hazzik** [comment]
I think the solution would be to `throw PlatformNotSupportedExeption`. Thoughts?

**Comment by fredericDelaporte** [comment]
Well, that was somewhat my first thought. But since we have avoided this in the call context case for feature parity sake, it would have been better to find a solution for web context too.

But it will be still better to throw that error than failing as it does currently.

**Comment by hazzik** [comment]
I don't think we would be able to find a viable solution at the moment.

**Comment by fredericDelaporte** [comment]
> I think the solution would be to throw `PlatformNotSupportedException` on access to WebSessionContext.

Thinking about it again, that may not be straightforward to do. For `netcoreapp` target, this is easy. But for `netstandard`, it is another story, since we do not really now what will be the executing runtime at compile time. It could even be the full framework after all.

Currently, there is no standardized way of detecting this. This is a work in progress in dotnet/corefx#17452.
Some workarounds are provided in dotnet/corefx#22293, like using [this](https://github.com/dotnet/corefx/blob/v2.1-preview1/src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.cs#L23-L25).

But maybe (even if unlikely) some other runtime than the full .Net Framework may support the `HttpContext`. So overall we may have better keeping this context working or failing as it does currently.

**Comment by hazzik** [comment]
I think the simplest fix would be to pass `throwOnError: true` into `System.Type.GetType` method in `ReflectiveHttpContext`.

**Comment by fredericDelaporte** [comment]
> I think the simplest fix would be to pass `throwOnError: true` into `System.Type.GetType` method in `ReflectiveHttpContext`.

While the code would be more correct this way (avoiding a potential null reference exception), it will not change anything to the case reported here, because that is a case which already throws even if `throwOnError` is `false`.

**Comment by danilobreda** [comment]
A "deprecated" or something like that would be nice, because even on .net full, the use of a DI is suggestive for a right way that the plataform is supporting right now.

**Comment by shatl** [comment]
@leandro86 Using DI instead would be a better approach.
I had a lot of issues with HttpContext based approach when decided to move some code to run in background thread.

**Comment by leandroaraujo** [comment]
@jpenniman, @shatl 
> In our ASP.Net Core apps, we don't use the WebSessionContext at all. We manage the ISessionFactory and ISession life cycle with the dependency injection built into ASP.Net Core...

And about the CloseSession?
.Net Core DI disposes/close Session automatically or I need to close by my own?

**Comment by maca88** [comment]
.Net Core DI disposes/close Session automatically or I need to close by my own?

The session will be closed and disposed automatically when the service provider scope is disposed. .Net Core DI [disposes](https://github.com/aspnet/DependencyInjection/blob/94b9cc9ace032f838e068702cc70ce57cc883bc7/src/DI.Specification.Tests/DependencyInjectionSpecificationTests.cs#L398) all services that implements `IDispose` interface when the scope that contains them is disposed. As NHibernate session implements `IDispose`, it will be disposed when the scope ends and also [closed if it not already closed](https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/Impl/SessionImpl.cs#L1535).

**Comment by lissinho** [comment]
> > .Net Core DI disposes/close Session automatically or I need to close by my own?
> 
> The session will be closed and disposed automatically when the service provider scope is disposed. .Net Core DI [disposes](https://github.com/aspnet/DependencyInjection/blob/94b9cc9ace032f838e068702cc70ce57cc883bc7/src/DI.Specification.Tests/DependencyInjectionSpecificationTests.cs#L398) all services that implements `IDisposable` interface when the scope that contains them is disposed. As NHibernate session implements `IDisposable`, it will be disposed when the scope ends and also [closed if it not already closed](https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/Impl/SessionImpl.cs#L1535).

We are using this approach but having trouble with ajax requests (odd "A command is already in progress" messages) and even "Session is closed!" sometimes.
Does anyone also have these problems? @jpenniman , @maca88 ?

**Comment by jpenniman** [comment]
It sounds like your scoping is incorrect. Make sure `ISessionFactory` is a singleton, and `ISession` is request scoped. You'll get a "A command is already in progress"  error if a session is running a sql command and you try to run another command on the same session.

```cs
services.AddSingleton<ISessionFactory>(...);
services.AddScoped<ISession>(...);
```
If you want to post your registrations here, I'll review them for any possible issues. If you don't want to post them publicly, you can email them to me.

We have ASP.net core applications using NHibernate handling thousands of transactions per second with no issues.

**Comment by lissinho** [comment]
> It sounds like your scoping is incorrect. Make sure `ISessionFactory` is a singleton, and `ISession` is request scoped. You'll get a "A command is already in progress" error if a session is running a sql command and you try to run another command on the same session.
> 
> ```cs
> services.AddSingleton<ISessionFactory>(...);
> services.AddScoped<ISession>(...);
> ```
> 
> If you want to post your registrations here, I'll review them for any possible issues. If you don't want to post them publicly, you can email them to me.
> 
> We have ASP.net core applications using NHibernate handling thousands of transactions per second with no issues.

Thanks for answering, @jpenniman 
Yeah, sounds like that. But we are registering just like you said. Here is the code:
```
            services.AddSingleton<ISessionFactory>((provider) => {
                var cfg = provider.GetService<ConfigManager>();
                return cfg.CreateSessionFactory();
            });

            services.AddScoped<NHibernate.ISession>((provider) => {
                var factory = provider.GetService<ISessionFactory>();
                return factory.OpenSession();
            });
```
ConfigManager is our class that, like it's name says, manage configuration. Here is CreateSessionFactory function code:
```
        public ISessionFactory CreateSessionFactory()
        {
            var dbc = Licenca?.dbConfigs.FirstOrDefault();

            if (dbc == null)
                throw new Exception("String de conexão não configurada !");

            IPersistenceConfigurer cfg;
            if (dbc.providerName == "OracleClient")
                cfg = OracleClientConfiguration.Oracle10
                                               .ConnectionString(c => c.Is(dbc.connectionString))
                                               .AdoNetBatchSize(0);
            else if (dbc.providerName == "System.Data.SqlClient")
            {
                cfg = MsSqlConfiguration.MsSql2008
                                        .ConnectionString(c => c.Is(dbc.connectionString))
                                        .AdoNetBatchSize(0);
            }
            else if (dbc.providerName == "Npgsql")
            {
                cfg = PostgreSQLConfiguration.PostgreSQL82
                                             .ConnectionString(c => c.Is(dbc.connectionString));
            }
            else
            {
                throw new Exception("Provider '" + dbc.providerName + "' não implementado !");
            }

            return Fluently.Configure()
                           .Database(cfg)
                           .Mappings(m =>
                                     m.FluentMappings.AddFromAssemblyOf<Entities.Classe>())
                           .BuildSessionFactory();
        }

```
We are using FluentNHibernate. Could you share your NH configuration?
Thanks again.

**Comment by lissinho** [comment]
To give more information:
We inject ISession in a separate BLL project as services, and inject services in controllers. Debugging, I realized that the line "factory.OpenSession" is called several times, not once per request. Is it correct?

**Comment by jpenniman** [comment]
Make sure your service classes in your BLL are also being registered in request scope.  If you make them singletons, then you've effectively made ISession a singleton (sort of... per service).

I assume your ConfigManager is also a singlton? (It should be, since ISessionFactory is a singleton.)

How are you debugging? If you're setting a breakpoint and using the browser to initiate the call, you'll see multiple calls while debugging. The browser retries the request for you automatically, up to 3 times. When not debugging, you should only get one ISession instance per request.

We don't use FluentNHibernate, but it shouldn't matter, unless Fluent is doing some magic that NHibernate-Core does not.

Here's how we configure NHibernate with Asp.Net Core. Note that we do **_not_** use WebSession. I abbreviated the OrderService implementation for simplicity/clarity by removing routing, mappers, logging, etc.:

``` cs
services.AddSingleton<Configuration>((provider) => {
    Configuration config = new Configuration();
           
    IDictionary<string, string> properties = new Dictionary<string, string> {
        { "connection.connection_string", this.Configuration.GetConnectionString("Default") },
        { "connection.driver_class", "MilestoneTG.NHibernate.TransientFaultHandling.SqlServer.SqlAzureClientDriver, MilestoneTG.NHibernate.TransientFaultHandling.SqlServer" },
        { "dialect", "NHibernate.Dialect.MsSql2012Dialect" },
        { "hbm2ddl.keywords", "auto-quote"},
        { "show_sql", "true"},
        { "generate_statistics", "true"}
    };

    config.AddProperties(properties);
    config.AddXmlFile("hibernate.hbm.xml");
});

services.AddSingleton<ISessionFactory>((provider) => {
    return provider.GetService<Configuration>().BuildSessionFactory();
});

services.AddScoped<ISession>((provider) => {
    ISessionFactory factory = provider.GetService<ISessionFactory>();
    ISession session = factory.OpenSession();
});

services.AddScoped<IOrderService, OrderService>();
```

Our `OrderService` takes in the `ISession`:

``` cs
public class OrderService : IOrderService
{
    ISession session;
    public OrderService(ISession session)
    {
        this.session = session;
    }

    public Task<Order> FindByIdAsync(long id)
    {
        return session.GetAsync<Order>(id);
    }

    public async Task<Order> Create(Order order)
    {
        var tran = session.BeginTransaction();
        try
        {
            await session.SaveAsync(order);
            tran.Commit();
            return order;
        }
        catch(Exception)
        {
            tran.Rollback();
            throw;
        }
    }
}
```

Our controller then takes in the `IOrderService`:

``` cs
public class OrderController : ControllerBase
{
    IOrderService service;
    public OrderService(IOrderService service)
    {
        this.service = service;
    }

    public Task<IActionResult> Get(long id)
    {
        return service.GetAsync(id);
    }
}

```

For your random "Session is closed" error, make sure ISession isn't wrapped in a `using` block any where, or that any code isn't explicitly closing or disposing the session.

**Comment by lissinho** [comment]
Thanks again, @jpenniman.
Yes, we are registering just like you've assumed: 
- services and session in request scope and confmanager and sessionFacory singleton;
- in services, we are injecting Session in constructor;
- in controllers, services are being injected in constructor;
- and we are not using `using` block in anyone of the codes that had triggered the random "Session is closed" error
So, everything seems to be exactly like you do. I am really lost here.

Does anyone have another suggestions?

**Comment by jpenniman** [comment]
If your configuration is correct, then there must be something with your business logic. If you want to email me directly, I can work with your company to do a formal architecture and code review: jason.penniman@milestonetg.com

There is likely something in your code that is trying to use ISession to make concurrent database calls under certain conditions, whether intentionally or unintentionally.

A few more considerations for DI:

* Make sure you're not creating any new scopes (`provider.CreateScope()`).
* Make sure you're not using service location anywhere other than startup and not trying to resolve scoped services from ApplicationServices.
* Make sure you're not injecting scoped or transient objects into the constructor of any middleware. Middleware instances are singletons.

Another issue I see a lot is improper use of threading patterns, Task and async/await. This can result in undesired and unpredictable behavior--such as concurrent calls to the same ISession instance and disposing of the scope before background work is finished.

* Async code mixed with synchronous code should be awaited. 
* Don't use any "fire and forget" patterns.
* Async scopes should be handled by the framework; don't create your own async scopes. 
* Similarly, don't try and make ISession or your service references AsyncLocal
* Make sure there are no async/void methods; all async method should return Task. 
* Make sure you are not trying to call async methods synchronously (calling .Result or .GetAwaiter().GetResult()).
* NHibernate supports async methods since 5.0. Microsoft recommends "async all the way" from your controllers down to your data access code.

Beyond that, it's difficult to know what is going on without looking at your code and design. At this point, I don't *_think_* it is an issue with NHibernate. ISession is not thread safe, and the same instance of ISession is being called concurrently by multiple threads in your code based on the errors you have described.

**Comment by lissinho** [comment]
Thanks again, @jpenniman .
I've created a project from "web application mvc" template and put my configuration code to reproduce the error here:
https://github.com/lissinho/netcorenhibernateajaxerror

You will see that it happens without any complex code (the routines just get datetime from database server), just configure a valid connection string to a postgresql server, run the project and keep refreshing the first page until see the errors.

I really don't know if it is a fluent, nhibernate, .net core DI or npgsql error. But I think it is an important issue to report.

I will keep trying to solve here, if I get any news, I'll let you know.

**Comment by jpenniman** [comment]
Your service is a static variable... there for a singleton... static instances exist only once per type.,,

``` cs
namespace WebApplication5.Controllers
{
    public class HomeController : Controller
    {
        private static IServiceTeste _serviceTeste;

        public HomeController(IServiceTeste serviceTeste)
        {
            _serviceTeste = serviceTeste;
        }
```

It must be an instance variable.  It should look like this:

``` cs
namespace WebApplication5.Controllers
{
    public class HomeController : Controller
    {
        private IServiceTeste _serviceTeste;

        public HomeController(IServiceTeste serviceTeste)
        {
            _serviceTeste = serviceTeste;
        }

```

**Comment by lissinho** [comment]
Perfect, @jpenniman !!!!! That's exactly what cause the errors!!!
Thanks a lot, man!!! I owe you a beer!!! 
XD

**Comment by jpenniman** [comment]
You're welcome. I'm glad it was something simple.

**Comment by gusgorman** [comment]
Personally I think having the ISession in the DI container is an anti pattern, because you might not want to rewire the whole application ever time you create a new hibernate session. Its much better for the ISessionFactory to live in the container..

For example, in my current application I am reusing the DAL (more specifically, the DAOs) in some batch processes that do some data migration. Here I don't want to re-create all the DAOs every time I create a new NHibernate session. 

In my current design each DAO has the ISessionFactory as a constructor dependency. If the dependency was changed to be an ISession this would cause a problem.

**Comment by jpenniman** [comment]
@gusgorman, If you look at my example, the whole application is not being wired-up every time an `ISession` is requested. `ISessionFactory` is managed by the container as a singleton and is only built once.

I agree that you may not always want to resolve `ISession` from the container and some scenarios may warrant injecting `ISessionFactory` instead, but that does not make this an anti-pattern.  There are exceptions to every rule. I cover this in my blog article on this topic: [https://jpenniman.blogspot.com/2019/07/using-nhibernate-in-aspnet-core.html](https://jpenniman.blogspot.com/2019/07/using-nhibernate-in-aspnet-core.html)

> If you do have a service the depends on NHibernate that should be a singleton, for example an in-memory cache or other service with an expensive instantiation, inject ISessionFactory instead, and use it to create a new session within the service itself.

In a typical service/application following [DDD](https://dddcommunity.org/) and/or [Clean Code/Architecture](https://blog.cleancoder.com/uncle-bob/2011/11/22/Clean-Architecture.html) principles, injecting `ISession` is generally appropriate, so that you are using the same session for the entire request as originally intended by the legacy `WebSessionContext`. Managing `ISession` inside each DAO can cause issues with lazy loaded associations. In DDD/Clean Code, your DAO should know nothing of your DTO or Mapper.  In that scenario, you may no longer have your `ISession` available when the mapper walks the graph. 

The other thing to be careful of when injecting `ISessionFactory`, as in your scenario, is you are now creating a separate database connection for each DAO needed for a given request. This can exhaust connection pools and connection limits (Oracle imposes hard limits on the number of connections), increase memory utilization on the database server, and have a negative performance impact on the application itself (creating connections is very expensive). I'm not saying your design has these issues, but just something to be concious of. In a typical use-case, injecting `ISessionFactory` instead of `ISession` would have all these issues. Again, there are always exceptions.

**Comment by gusgorman** [comment]
@jpenniman  Most of your reply is a straw man. I am not advocating managing the hibernate session inside each DAO, so there is no danger of lazy loading errors or creating a database connection for each DAO. This would be another anti pattern and I'm not sure why you have to come to conclusion that I am doing this, because I certainly didn't say anything to indicate that I am (I know you slipped the caveat in at the end say to you're not sure if my design have these issues, but I really don't think any of it was worth mentioning at all).

The normal and correct use of DAOs is of course to inject them into the Service layer (or other higher layer of the application). 

In the system I am currently working on we have one hibernate session per Http Request for the web app, and one hibernate session per "Unit of Work/Transaction" in the data migration batch processes. 

I have had a quick look at your blog post and interestingly you say this:

> "A common mistake I see among developers new to dependency injection (and even some veterans) is registering the services as singletons. This causes undesired behavior because you’ve effectively made ISession a singleton as well, since it will never be instantiated again for furture requests"

This is where we disagree. I would say the mistake is injecting the ISession into anything at all, so that it doesn't dictate the scope of anything that needs to consume it. It isn't good idea thing to tie the lifetime of any component to the lifetime of a hibernate session. Injecting the ISessionFactory is much more flexible.

**Comment by jpenniman** [comment]
@gusgorman, You make a very interesting point. You mentioned that you are injecting ISessionFactory into the DAO:

> In my current design each DAO has the ISessionFactory as a constructor dependency.

I made the assumption that you would have to call ISessionFactory.OpenSession() within the DAO which would create a new session/connection in every DAO, and I should not have assumed your usage; I apologize.

Can I ask, then, if you are injecting ISessionFactory into the DAO, how are you managing ISession? I am curious how your design works. I'm always open to learning different ways of tackling a problem. 

Can you share a web app example, since this thread is about Asp.Net Core and the lack of a WebSessionContext? It would help me and others reading the thread to understand how to best use your approach. 

I agree, your approach gives much finer grained/flexible control over session lifetime. I can update my article to include your approach (crediting you, of course).

**Comment by gusgorman** [comment]
@jpenniman Apology 100% accepted!

tbh I don't have the time to prepare a full app to share at the moment, but my current design is really no different from that which was encouraged by Castle and Spring.NET 10 years ago.... and Java Spring and similar frameworks before that.

So the layers in a web app (from top to bottom):

1) MVC (or Razor pages) - Controllers have Services injected into them
2) Service layer.  Transactions are managed here using AOP (we use a Transaction attribute which signals the DI container to wrap the service method in an interceptor).  The code in the services has no reference to or knowledge of NHibernate whatsover. DAOs are injected into the services.
3) Data Access Layer consisting of DAOs. The DAOs have the ISessionFactory injected into them. Data access logic and Nhibernate specific code are encapsulated within the DAOs.

We have one NHIbernate session per http request, which is created and disposed using IActionFilter and IResultFilter implementations. This is similar to the "OpenSessionInView" pattern used in old Spring apps. which some people considered an anti-pattern when it was originally used :) However, I find it works well (we only ever pass ViewModels into the views so the views are decoupled from the actual domain objects).

The only small hack in the current web app is that we are using a modified version of the AspNetCoreWebSessionContext  suggested by leandro86 earlier in this thread. The only difference  being that the static method to set the HttpAccessor looks like this:

```
        public static Func<IHttpContextAccessor> ContextAccessorInitialiser
        {
            set { _httpContextAccessor = new Lazy<IHttpContextAccessor>(value); }
        }
```
This is probably the only static property in our current architecture and the moment, which is a small shame but not the worse hack in the world.

The batch process have a similar architecture, the main difference being that the top layer is some C# code that compiles to an exe, and we are using a "Session per transaction" implementation of our Transaction interceptor, ie the interceptor implementation creates the Hibernate Session, and opens a transaction before invoking the wrapped service call, and then commits the transaction and closes then session after the call.  This is flexible though - for example you can make sure your interceptor implementaion checks to see if a transaction is already running if you want to wrap multiple service calls in a single transaction or ISession.

I hope this makes sense! If you've got any questions fire away, or I'll try to dig up some code examples when I can find the time.

**Comment by jpenniman** [comment]
@gusgorman, Ah, that was the missing piece... you have a custom WebSessionContext to manage session lifetime.  Then you can rely on ISessionFactory.GetCurrentSession() to work correctly in both scenarios. That all makes perfect sense.

Yeah, some regard using ActionFilters for working with the NHibernate session as an anti-pattern because there are scenarios where an ActionFilter can be called more than once for a single request. For example, if you have an action that returns the results of another action method, which I think is a bad practice to begin with. The other way to handle that would be with a custom middleware and make it part of the request pipeline the you're guaranteed to only be called once. But, if it's working for you, great!

It would be nice to get something "official" into NHibernate itself (or an extension to) now that the full framework has been end-of-lifed (but will continue to get security updates for at least another 10 years).  More and more companies are moving to .Net Core on Linux as part of their cloud migrations.

**Comment by felipeoriani** [comment]
Hey guys. We have a huge project made in `.Net Framework` here where we use `NHibernate` as persistence layer. I was trying to implement a POC for a new version using `Asp.Net Core` with `NHibernate`. We have implemented the `session per request` on our current application and I was trying to find a way to implement a Middleware to do the same behaviour on the `ASP.NET Core` and after a lot of problems I found this thread. I would love to see some way to bind a session on some place and access it from `SessionFactory.GetCurrentSession()` but it seems to be not implemented/supported on `.Net Core`. I know a lot of people want it on `.Net Core`.

Is there any plans from NHIbernate Team to implement it? While it is not ready, what is a good approach to workaround on it (`session/transaction per request`)?

**Comment by danilobreda** [comment]
@felipeoriani Using DI instead would be a better approach. 
The idea of "scoped" on DI, is a session/transaction per request on asp net core, for example.
Here's a example of a DI Setup using SimpleInjector but you can use the DI that comes with aspnetcore.

A singleton sessionfactory.
`kernel.RegisterSingleton<ISessionFactory>(() => MySessionFactoryProvider.CreateSessionFactory());`
            
A scoped "ISession"
`kernel.Register<IUnitOfWork, UnitOfWork>(Lifestyle.Scoped);` (injection of isessionfactory on constructor)

thats only a example, on how you inject the isessionfactory and isession

I still remain confident with my [comment ](https://github.com/nhibernate/nhibernate-core/issues/1632#issuecomment-394209988)from almost 2 years ago.
> 
> 
> A "deprecated" or something like that would be nice, because even on .net full, the use of a DI is suggestive for a right way that the plataform is supporting right now.

**Comment by felipeoriani** [comment]
@danilobreda, yes I have been trying some ways to manage the `ISession` and `ITransaction` in a asp.net core web application and yes, it is the best way. I used to bind the `ISession` using the CurrentSessionContext but it is not available anymore because of the platform like I understand from this topic.

If we add the `ISession` from the `AddScoped` registry method on the `ISessionCollection`, I know it will just open the session when it is necessary to inject on any dependency and it's fine. But using Oracle, I have to clear the pools and close the Session. I have defined a `Middleware` to to do. Not sure if it is good but works fine. 
```
public async Task InvokeAsync(HttpContext httpContext)
{
   await _next.Invoke(httpContext);

   ISession session = httpContext.RequestServices.GetService<ISession>();
   if (session != null)
   {
       ClearPool(session.Connection);

       if (session.IsOpen)
          session.Close();
               
    }
}
```
If someone looks how to close an ISession, it's working fine (for async/non-async requests).

**Comment by erizzo** [comment]
@jpenniman I have 2 questions about the DI managed pattern that you describe:

1. Do you open a transaction when creating the session? In JavaI would use the Spring annotations for declarative transaction boundaries, but I don't think .Net has the equivalent. And I don't like individual services or controllers having to explicitly manage transactions all the time (sometimes yes, but those are exceptions).
On a related note, how does the transaction get rolled back if an exception occurs during the request? Is explicitly calling `session.Transaction.Rollback()` not necessary?
2. What about things that aren't subject to DI injection? I'm thinking specifically of controller attributes (action filters) that might need the session.

**Comment by jpenniman** [comment]
@erizzo, good questions. I'll try and answer as best I can...

> 1. Do you open a transaction when creating the session?

That depends on the scenario. Personally, I always manage explicit transactions and do so as close to the database work as possible to shorten transaction times and reduce the likelihood of deadlocks or other performance issues related to long transactions. This is exactly what you would be doing with the `@Transaction` attribute in Spring Data JPA, but as you said, there is no .Net equivalent. (You could always recreate the behavior with PostSharp).

Let's look at some options...

1) If you do nothing, NHibernate will fallback to implicit transactions. A transaction will be created and committed (or rolled back) automatically every time you call a persistence method such as `ISession.Save()`. This is not recommended and is [discouraged in the docs](https://nhibernate.info/doc/nhibernate-reference/transactions.html):

> We advise against such a pattern, because this causes each single interaction of the session with the database to be a transaction on its own. This causes overhead, breaks the unit of work in case of errors during flushing, and causes some features to be disabled like auto-flush and second level cache.

2) Explicitly begin the transaction at the start of a request, and commit at the end of the request.  You could handle this with a custom middleware. Rollback would be handled in a custom exception middleware/filter. However, this means that exceptions would have to bubble-up all the way to the middleware/global exception handler. This may not always be desirable. This approach also leaves a connection and transaction open for the entire request. This is a code smell even without NHibernate. It also means you end up with a session even when you don't need one.

3) [Wrap `ISession` in a custom UnitOfWork and Repository](https://nhibernate.info/doc/patternsandpractices/nhibernate-and-the-unit-of-work-pattern). I see this a lot. I personally find this approach redundant as `ISession` **_is_**  an implementation of the Repository Pattern and UnitOfWork Pattern. However, this can add advantages in unit testing as well as an abstraction around transactions and when they happen. 

4) Do what I do and begin and end the transaction as close to the work as you can. Since I typically don't use repository classes when using NHibernate (since ISession is a repository) I do it in my service class.  As I mentioned earlier, you could use PostSharp to put that boilerplate code in an aspect.



> On a related note, how does the transaction get rolled back if an exception occurs during the request? Is explicitly calling session.Transaction.Rollback() not necessary?

Anything that has not been committed with rollback when the session is closed.



> What about things that aren't subject to DI injection? I'm thinking specifically of controller attributes (action filters) that might need the session.

We have access to the container in ActionFilters in the context. So while they do no support constructor injection, you can use service location...

``` cs
public override void OnActionExecuting(ActionExecutingContext context)
{
    ISession session  = context.HttpContext.RequestServices.GetService<ISession>();
}
```
[Cusom middleware does support constructor injection](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/write?view=aspnetcore-3.1#middleware-dependencies) and is the recommended way of providing singleton dependencies to middleware:

> Middleware should follow the Explicit Dependencies Principle by exposing its dependencies in its constructor. Middleware is constructed once per application lifetime.

...but because middleware is only constructed once, you can use service location to grab the ISession instance. Again, the container is available on the context:

``` cs
public async Task Invoke(HttpContext httpContext)
{
    ISession session  = httpContext.RequestServices.GetService<ISession>();
    await _next(httpContext);
}
```
... or [Microsoft's recommended approach of method injection](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/write?view=aspnetcore-3.1#per-request-middleware-dependencies), though this will only work with Microsoft's DI...

> Because middleware is constructed at app startup, not per-request, scoped lifetime services used by middleware constructors aren't shared with other dependency-injected types during each request. If you must share a scoped service between your middleware and other types, add these services to the Invoke method's signature. The Invoke method can accept additional parameters that are populated by DI

``` cs
public async Task Invoke(HttpContext httpContext, ISession session)
{
    await _next(httpContext);
}
```

I hope this helps. Cheers!

**Comment by felipeoriani** [comment]
Very nice answer @jpenniman,

I have been implementing the transaction from the controller scope with an actionFilter. We can access the container from the `context.RequestServices` and get the current `ISession`. 

I know the ISession will be closed, but we have an enterprise application that can run over Sql Server, Oracle and PostgreSql. In the case of Oracle, I have to clear the pools from the DbConnection inside the ISession to avoid problems with the next connections. We created a custom middleware and just call the `next` and then give the correct treatment to the ISession. 
```
public async Task InvokeAsync(HttpContext httpContext)
{
    await _next.Invoke(httpContext);

    ISession session = httpContext.RequestServices.GetService<ISession>();

    if (session != null)
    {
        ClearPool(session.Connection);

        if (session.IsOpen)
            session.Close();                
    }
}
```
I am not sure if it is a good pratice to create a middleware for it, but that's the solution I got. If you think there is a better way, please share with us.

Thank you!

**Comment by jpenniman** [comment]
Thanks @felipeoriani.  It's interesting that you have to clear the connection pool in Oracle. I haven't run across the need for it before. I used Oracle quite a bit with full framework, but not much with corefx (mostly because of poor driver support from Oracle). If you want to eliminate the need, you could try a different provider. Or open a ticket with Oracle. It sounds more like a connection pool bug than a NHibernate thing.

As far as best practice, I would say, if your solution works, isn't causing any performance issues, and you can test your code easily, then stick with it, and document it well, both in the code and out. The only downside I can see is it's always creating a session regardless if one is needed.

**Comment by pastidev** [comment]
@jpenniman thank you to give some light about this topic, i've tried to understand why people keeps trying use WebSessionContext and check the binding on the CurrentSessionContext in examples like NHibernate and can't get it, because like you say if you use DI (requestScope) of ISession you get the same session for all the request.

Now i'm using self hosted owin with autofac trying to emulate an old example made for wep api with ninject and hibernate. If i use autofac and check the activation of a ISession only gets activated once, but the same using the original webapi2 and ninject (no owin) is activated each time a ISession is requested. I think that's because they used the check on CurrentSessionContext to Bind a new session or not. But i don't understand why with autofac works without that.

  ```
  container.Bind<ISessionFactory>().ToConstant(sessionFactory);
            container.Bind<ISession>().ToMethod(CreateSession).InRequestScope();
 private ISession CreateSession(IContext context)
        {
            var sessionFactory = context.Kernel.Get<ISessionFactory>();
            if (!CurrentSessionContext.HasBind(sessionFactory))
            {
                var session = sessionFactory.OpenSession();
                CurrentSessionContext.Bind(session);
            }
            return sessionFactory.GetCurrentSession();
        }
```

**Comment by marciliooliveira** [comment]
> Very nice answer @jpenniman,
> 
> I have been implementing the transaction from the controller scope with an actionFilter. We can access the container from the `context.RequestServices` and get the current `ISession`.
> 
> I know the ISession will be closed, but we have an enterprise application that can run over Sql Server, Oracle and PostgreSql. In the case of Oracle, I have to clear the pools from the DbConnection inside the ISession to avoid problems with the next connections. We created a custom middleware and just call the `next` and then give the correct treatment to the ISession.
> 
> ```
> public async Task InvokeAsync(HttpContext httpContext)
> {
>     await _next.Invoke(httpContext);
> 
>     ISession session = httpContext.RequestServices.GetService<ISession>();
> 
>     if (session != null)
>     {
>         ClearPool(session.Connection);
> 
>         if (session.IsOpen)
>             session.Close();                
>     }
> }
> ```
> 
> I am not sure if it is a good pratice to create a middleware for it, but that's the solution I got. If you think there is a better way, please share with us.
> 
> Thank you!

Hi @felipeoriani and @jpenniman , would you have any idea how we could bind the session when trying to load a related entity from a custom converter using automapper? I've been facing a problem with that. Since the session is created at any time when the dependency is necessary, when I resolved the entity using a custom converter on automapper class it is in a different session, so if try to save it .. 

I getting an error like "[HibernateException: illegally attempted to associate a proxy with two open Sessions]" since we can't get a currentsessioncontext from the .net core framework.

I appreciate it!

**Comment by jpenniman** [comment]
@marciliooliveira , if the instance of ISession is request scoped, `AddScoped()`, you shouldn't be getting a new session in your automapper. Could it be you're registering it transient scope (`AddTransient()`)?

I personally don't use automappers for my DTOs, so I can't speak to it from experience.

**Comment by marciliooliveira** [comment]
@jpenniman thank you for the answer.
The ISession instance is requested scoped, but when I try to get an instance of some repository from the current provider (ServiceCollection) in order to inject it into the custom converter, I get a null and I have to register it in a new provider scope and it is when the new ISession come up. My repositories are registered as transient.

Something like that.. just simplified for understanding 

```
var config = new MapperConfiguration(config =>
            {
                    config.ConstructServicesUsing(type => {

                    var entityType = type.GetGenericArguments()[0];
                    var repositoryType = typeof(InterfaceMyRepository<>).MakeGenericType(entityType);
                    var repository = services.BuildServiceProvider()
                                             .GetRequiredService(repositoryType);  // Always NULL if I don't registered it previously
                                                                                                          // If I registered it services.AddTransient(repositoryType) 
                                                                                                          // A new ISession is opened

                    ActivatorUtilities.CreateInstance(provider, type, repository);
            }
});
```
That configuration is in an extension called on the Configure(...) method on the Program.cs

But anyway, thanks a lot for answering that.

**Comment by jpenniman** [comment]
There is your problem. You're building 2 containers.

**Comment by jpenniman** [comment]
@marciliooliveira ,if you want to put together a small,complete example reproduction in your repository, I'll take a look.

**Comment by marciliooliveira** [comment]
@jpenniman I'll do that and let you know as soon as I finish it.
@felipeoriani I agree and I tried but when we have some generic repository needed to be injected on custom converter the DI doesn't resolved it, but yes the main issue is how to inject on automapper scopes as you said. I'll ask on community. Thank you!

**Comment by amod81** [comment]
@jpenniman 
In one of your posts you said you don't use the repository layer and I see in that e.g code you have the session injected in the service code itself, however I have a repository layer and I want my service layer to function as single UnitOfWork. In order to achieve that none of the repository code has transaction related actions commit/rollback etc. Since the ISession is scoped per request, I inject it in service and the repository as well. When my service gets a call from the mediator/controller, i begin the transaction, invoke multiple repositories (select/insert/update) etc and at the end commit the transaction in the service. Should there be any problem with this approach?

**Comment by jpenniman** [comment]
That is fine. It is the precise reason we want the `ISession` scoped to the request--so we get the same session instance anytime we resolve it throughout the request.

If you want to further abstract things away, you could create a `IUnitOfWork` and implementation that takes in ISession and exposes BeginTransaction, Commit, Rollback. This will give you full mocking capability for testing and completely hide NHibernate from your business logic.

A side note on my repository comment earlier... I recently ran into a situation where a system we built for a customer for internal use, the customer now wants to sell it on the commercial market (closed source). NHibernate is LGPL and cannot be used. I'm wishing I had buried NHibernate in repositories and a unit of work.

**Comment by marciliooliveira** [comment]
Hi @jpenniman @felipeoriani 
Jimmy (from AutoMapper) answered me and just advised to NOT using the ConstructUsingServices automapper funcionallity 'cause the DI package handles the injection into the custom converter. So we just need to ensure registering all custom ITypeConverter types and generic repositories as well as other dependents classes. Now, as I'm not creating a new containers, the ISession is only one during the request scope. As far as I've tested, everything went well.
Thanks for helping.

**Comment by chengang-hub** [comment]
Hi @jpenniman Hope to get your help，After I migrated my project to NetCore, I found a possible memory leak. 
Nhibernet consumes 500M memory when initializing：
configuration.BuildSessionFactory() : 300M
MappingByCode: 80M
![image](https://user-images.githubusercontent.com/108459883/216240143-f52138ee-0272-467e-a82b-a059e2e31c14.png)

After a day, the system occupies 5G memory
![image](https://user-images.githubusercontent.com/108459883/216240177-34074987-25d5-451b-9f5b-6a76df5c9e3c.png)

this is my code，Is it my usage wrong？ thanks ：

       IServiceCollection _services;
        public void ConfigureServices(IServiceCollection services)
        {
            _services = services;
            ParseEnvConfig();
            LocalSetting.Init(Configuration.GetSection("YCH"));
            #region AddMvc

            services.AddMvc(options => { options.EnableEndpointRouting = false; });
            services.AddMvc(op => {
                op.Filters.Add(new ActionAuthFilter());
                op.Filters.Add(new ActionLogFilter());
                op.Filters.Add(new CustomExceptionFilter());
                op.Filters.Add(new CustomActionFilter());
            }).AddNewtonsoftJson(options =>
                {
                    options.SerializerSettings.DateFormatString = "yyyy-MM-dd HH:mm:ss";
                    options.SerializerSettings.ContractResolver = new DefaultContractResolver();
                    options.SerializerSettings.Converters.Add(new GuidConverter());
                    options.SerializerSettings.NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore;
                })
            .ConfigureApplicationPartManager(m => {
                var feature = new ControllerFeature();
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.Game")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.GameMachine")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.Game.Investment")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.Game.PackageTicketOTA")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.Game.PhysicalCoin")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.Game.Race")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.Game.Sports")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.Game.Lease")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.SpecialActivity")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.ManageOnline")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.MicroReport")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.OnLine")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.ShoppingGuide")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.TikTok")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.Ykb")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.CommonInterface")));
                m.ApplicationParts.Add(new AssemblyPart(Assembly.Load("Mall.Services.PrivateInterface")));
                m.PopulateFeature(feature);
                services.AddSingleton(feature.Controllers.Select(t => t.AsType()).ToArray());
            });
            services.Configure<IISOptions>(options => {
                options.ForwardClientCertificate = false;
            });
            #endregion

            #region Register
            services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
            services.AddSingleton<ISessionService, SessionService>();
            ConfigNHibernate(services);
            ExecSetting(services);
            services.BuildServiceProvider().GetServices<IBLLInfo>().OrderByDescending(a => a.IsMaster).Foreach(info => info.Register(services)); ;

            InitServerDirs();
            #endregion

            #region AddSession

            services.AddDistributedMemoryCache();
            services.AddSession(options =>
            {
                options.Cookie.SameSite = SameSiteMode.Unspecified; 
            }); 
            #endregion
        }


private void ConfigNHibernate(IServiceCollection services)
        {
            var connectionString = LocalSetting.Curr["ConnectionString"];
            NHibernate.Cfg.Configuration configuration = new NHibernate.Cfg.Configuration();
            configuration.Properties.Add("connection.provider", "NHibernate.Connection.DriverConnectionProvider");
            configuration.Properties.Add("connection.driver_class", "NHibernate.Driver.SqlClientDriver");
            configuration.Properties.Add("connection.connection_string", connectionString);
            configuration.Properties.Add("dialect", "NHibernate.Dialect.MsSql2008Dialect");
            configuration.Properties.Add("command_timeout", "300");

            configuration.Properties.Add("current_session_context_class", typeof(RequestSessionContext) + "," + typeof(RequestSessionContext).Assembly);
            services.AddSingleton(configuration);

            MappingNHibernate(configuration);
            try
            { 
                services.AddSingleton(typeof(NHibernate.ISessionFactory),configuration.BuildSessionFactory());
                services.AddScoped(provider => {
                    var factory = provider.GetService<NHibernate.ISessionFactory>();
                    if (factory == null)
                    {
                        throw new InvalidOperationException($"Can not get service {typeof(NHibernate.ISessionFactory)}");
                    }
                    return factory.OpenSession();
                }); 
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
            }
        }

private void MappingNHibernate(NHibernate.Cfg.Configuration configuration)
        { 
            var models = new List<string>(AA.DllLoader.SearchFile("Mall.Model.dll"));
            models.AddRange(
                AA.DllLoader.SearchFile("Mall.Model.*.dll")
            );   
            foreach (var model in models)
            {
                var assembly = AA.GetAssembly(model);
                NHibernate.Mapping.ByCode.NHConfig.MappingByCode(configuration, assembly);
            } 
        }

 public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            var sp = app.ApplicationServices;
            Mall.Core.HttpContext.Configure(sp.GetRequiredService<IHttpContextAccessor>());
            EnginContext.Initialize(new EngineFactory(_services, sp));
             。。。。。。
}


//This project was migrated from the old project and instead of controller injection, manual injection was used

    public class EngineFactory 
    {
        private readonly IServiceCollection _services;
        private IServiceProvider _serviceProvider;
        public EngineFactory(IServiceCollection services, IServiceProvider provider)
        {
            _services = services;
            _serviceProvider = provider;
        }

        static ThreadLocal<IServiceScope> ThreadServiceScope = new ThreadLocal<IServiceScope>();

        public void Close()
        {
            if(ThreadServiceScope.Value != null)
            {
                ThreadServiceScope.Value.Dispose();
                ThreadServiceScope.Value = null;
            }
        }



        public void AddSingleton<TService>() where TService : class
        {
            _services.AddSingleton<TService>();
        }
        public void AddSingleton<TService, TImplementation>() where TService : class where TImplementation : class, TService
        {
            _services.AddSingleton<TService, TImplementation>();
        }
        public void AddSingleton(Type serviceType, Type implementationType)
        {
            _services.AddSingleton(serviceType, implementationType);
        }
        public void AddSingleton(Type serviceType)
        {
            _services.AddSingleton(serviceType);
        }
 
        public object Resolve(Type type)
        {
            return _serviceProvider.GetService(type);
        }

        public T Resolve<T>()
        {
            return _serviceProvider.GetService<T>();
        }

        public T ResolveScope<T>()
        { 
            return ResolveOper(sp => sp.GetService<T>());
        }

        public List<T> ResolveList<T>()
        {
            return ResolveOper(sp => sp.GetServices<T>()).ToList();
        }
        public T ResolveNamed<T>(string serviceName)
        {
            return _serviceProvider.GetNamedService<T>(serviceName);
        }

        private T ResolveOper<T>(Func<IServiceProvider, T> func)
        {
            if (ThreadServiceScope.Value == null)
                ThreadServiceScope.Value = _serviceProvider.CreateScope();
            return func(ThreadServiceScope.Value.ServiceProvider);
        }
    }
    public class EnginContext
    {
        private static EngineFactory _engine; 

        [MethodImpl(MethodImplOptions.Synchronized)]
        public static EngineFactory Initialize(EngineFactory engine)
        {
            return _engine ?? (_engine = engine);
        } 
        public static EngineFactory Current => _engine;
    }


//this is NHibernate  current_session_context_class
public class RequestSessionContext : ICurrentSessionContext
    {
        ISessionFactoryImplementor CurrFac;
        public RequestSessionContext(ISessionFactoryImplementor factory)
        {
            CurrFac = factory;
        }
        public NHibernate.ISession CurrentSession()
        {
            return EnginContext.Current.ResolveScope<NHibernate.ISession>();
        }
    }


// and  tihs is  BLL query and Add 

public virtual void DBAdd(T model, object populateWith = null)
        {
            if (Licence.IsQueryOnly)
                return;
            SetDefualtValue(model);
            if (populateWith != null)
            {
                model.PopulateWith(populateWith);
            }
            Licence.CheckAdd(model);
            using (MallTrans tr = DBContext.BeginTran())
            {  
                NHSessionFac.CurrFac.GetCurrentSession().Save(model); 
                tr.Commit();
            }
        }


 public virtual IList<U> SelectBase<U>(
            Expression<Func<T, IProjection>>[] Selects,
            Expression<Func<T, Order>>[] Orders,
            PageInfo page,
            params Expression<Func<T, bool>>[] contitions)
        {
            ISession sess =NHSessionFac.CurrFac.GetCurrentSession();

            var fun = new Func<IList<U>>(() =>
            {
                IQueryOver<T, T> query = sess.QueryOver<T>();
                if ((typeof(Mall_LogBase)).IsAssignableFrom(typeof(T)))
                {
                    query = AddBLLContid(query);
                }
                else
                {
                    var syscondition = GetAddConditoin();
                    syscondition.ForEach(o => query.Where(o));
                }
                foreach (var ex in contitions)
                {
                    if (ex == null) continue;
                    query.Where(ex);
                }
                if (Selects != null)
                {
                    IProjection[] res = new IProjection[Selects.Length];
                    for (int i = 0; i < res.Length; i++)
                    {
                        res[i] = ExpressionToInvoke<IProjection>(Selects[i]);
                    }
                    query.Select(res);
                }
                if (Orders != null)
                {
                    for (int i = 0; i < Orders.Length; i++)
                    {
                        query.UnderlyingCriteria.AddOrder(ExpressionToInvoke<Order>(Orders[i]));
                    }
                }

                if (page != null)
                {
                    if (!page.NotGetTotle)
                    {
                        page.TotalCount = query.RowCount();
                    }
                    query.Take(page.Take).Skip(page.Skip);
                }
                return query.List<U>();
            });
           return fun();
        }



 public static MallTrans BeginTran()
        { 
            ISession session = NHSessionFac.CurrFac.GetCurrentSession();
            if (session.Transaction.IsActive)
            {
                return new MallTrans(null);
            }
            ITransaction tr = session.BeginTransaction();
            return new MallTrans(tr);
        }





    public class MallTrans : IDisposable
    {
        public MallTrans(ITransaction tr)
        {
            NHTransaction = tr;
        }

        public void Dispose()
        {
            RollBack();
            //if (NHTransaction != null)
            //{
            //    NHTransaction.Dispose();

            //}
        }

        /// <summary>
        /// 提交数据
        /// </summary>
        public void Commit()
        {
            if (NHTransaction != null)
            {
                NHTransaction.Commit(); 
                  NHSessionFac.CurrFac.GetCurrentSession().Clear(); 

            }
        }

        public void RollBack()
        {
            if (NHTransaction != null)
            {
                
                if ((!NHTransaction.WasRolledBack) && (!NHTransaction.WasCommitted))
                {
                    NHTransaction.Rollback();
                    NHSessionFac.CurrFac.GetCurrentSession().Clear();

                }
            }
        }

        public bool WasRolledBack
        {
            get
            {
                if (NHTransaction != null)
                {
                    return NHTransaction.WasRolledBack;
                }
                return false;
            }
        }

        public bool WasCommitted
        {
            get
            {
                if (NHTransaction != null)
                {
                    return NHTransaction.WasCommitted;
                }
                return false;
            }
        }

        ITransaction NHTransaction
        {
            get;
            set;
        }
    }

**Comment by Archer-firecracker** [comment]
@hazzik https://stackdev.space/questions/what-techniques-can-be-used-to-speed-up-insertion-processes-in-nhibernate
